src/D2L.CodeStyle.Analyzers/DependencyScope/DependencyScopeAnalyzer.cs

```diff
@@ -0,0 +1,142 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace D2L.CodeStyle.Analyzers.DependencyScope {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class DependencyScopeAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.SingletonDependencyIsWebRequest
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+			context.RegisterSyntaxNodeAction(
+				AnalyzeClass,
+				SyntaxKind.ClassDeclaration
+			);
+		}
+
+		private void AnalyzeClass( SyntaxNodeAnalysisContext context ) {
+			var root = context.Node as ClassDeclarationSyntax;
+			if( root == null ) {
+				return;
+			}
+
+			var symbol = context.SemanticModel.GetDeclaredSymbol( root );
+			if( symbol == null ) {
+				return;
+			}
+
+			// skip classes not marked singleton
+			if( !symbol.IsTypeMarkedSingleton() ) {
+				return;
+			}
+
+			AnalyzeType( context, root, symbol );
+		}
+
+		private void AnalyzeType( SyntaxNodeAnalysisContext context, ClassDeclarationSyntax root, ITypeSymbol type ) {
+			if( type.IsTypeMarkedWebRequest() ) {
+				var location = GetLocationOfClassIdentifierAndGenericParameters( root );
+				context.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.SingletonDependencyIsWebRequest,
+					location
+				) );
+			}
+
+			foreach( ISymbol member in type.GetExplicitNonStaticMembers() ) {

```

Comment:
I think you could not use this extension. It probably makes sense for static things too. I don't think you need to worry about implicitly declared members.

Reply:
i.e. I think you could just use `type.GetMembers()`